test/Microsoft.Repl.Tests/ShellTests.cs

```diff
@@ -575,5 +578,66 @@ public async Task RunAsync_WithCtrlEKeyPress_VerifyCaretPositionWasUpdated()
+
+            cancellationTokenSource = new CancellationTokenSource();
+            MockConsoleManager mockConsoleManager = new MockConsoleManager(consoleKeyInfo, cancellationTokenSource);
+
+            ShellState shellState = new ShellState(defaultCommandDispatcher,
+                consoleManager: mockConsoleManager);
+
+            return new Shell(shellState);
+        }
+
+        private static Shell CreateShell(ConsoleKeyInfo consoleKeyInfo, string previousCommand, string nextCommand, out CancellationTokenSource cancellationTokenSource)
+        {
+            return CreateShell(new ConsoleKeyInfo[] { consoleKeyInfo }, previousCommand, nextCommand, out cancellationTokenSource);
+        }
+
+        /// <summary>
+        /// Converts a string into the series of ConsoleKeyInfo instances that would be used
+        /// to type the string in the console.
+        /// </summary>
+        private static List<ConsoleKeyInfo> GetConsoleKeyInfo(string text)
+        {
+            List<ConsoleKeyInfo> keys = new();
+
+            text = text.Replace("\r\n", "\n");
+
+            foreach (char c in text)
+            {
+                ConsoleKeyInfo consoleKeyInfo = GetConsoleKeyInfo(c);
+                keys.Add(consoleKeyInfo);
+            }
+
+            return keys;
+        }
+
+        /// <summary>
+        /// Builds a ConsoleKeyInfo object with the specified console key, the null keyChar ('\0') and no modifiers.
+        /// Intended for non-printable keystrokes.
+        /// </summary>
+        private static ConsoleKeyInfo GetConsoleKeyInfo(ConsoleKey key) => new('\0', key, shift: false, alt: false, control: false);
+
+        /// <summary>
+        /// Builds a ConsoleKeyInfo that could produce the specified character
+        /// </summary>
+        private static ConsoleKeyInfo GetConsoleKeyInfo(char keyChar)
+        {
+            return keyChar switch
+            {
+                >= 'a' and <= 'z'        => new(keyChar: keyChar, key: (ConsoleKey)(keyChar - 32), shift: false, alt: false, control: false),
+                >= 'A' and <= 'Z'        => new(keyChar: keyChar, key: (ConsoleKey)keyChar,        shift: true,  alt: false, control: false),
+                >= '0' and <= '9' or ' ' => new(keyChar: keyChar, key: (ConsoleKey)keyChar,        shift: false, alt: false, control: false),
+                ':'                      => new(keyChar: keyChar, key: ConsoleKey.Oem1,            shift: true,  alt: false, control: false),
+                '/'                      => new(keyChar: keyChar, key: ConsoleKey.Oem2,            shift: false, alt: false, control: false),
+                '"'                      => new(keyChar: keyChar, key: ConsoleKey.Oem7,            shift: true,  alt: false, control: false),
+                '\n'                     => new(keyChar: '\r',    key: ConsoleKey.Enter,           shift: false, alt: false, control: false),
```

Comment:
I know its a little odd that `\n` produces a ConsoleKeyInfo instance with a KeyChar of `\r`, but that seems to match what you'd get if you hit the Enter key on your keyboard, so that's what I used.